### PR TITLE
feat: search entire UniswapV3 fee range for longtail swaps

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Adds UniswapV3 pool optimization to long-tail swaps, enabling more assets and better rates.

This PR adds the ability to search all fee ranges for a given asset pair's available pools, get a quote for each, and transact against the contract resulting in the best rate. In particular, we'll get better rates for assets with tight liquidity pools.

Also adds support for additional assets that did not have pools for the `3000` fee range of UniswapV3.

An example to see this working: compare the rate between this branch and prod for LINK to DOGE.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/5878

## Risk

High risk to long-tail swaps - we'll want to test these again, as messing up the contracts here (e.g. using a non-whitelisted  contract) could lead to funds landing in a black hole.

## Testing

Perform longtail swaps on a range of assets (the sell asset is important, not so much the buy asset - so just use the cheapest).

### Engineering

Sanity check the contract selection logic - using the wrong aggregator contract will cause lost funds.

### Operations

☝️

## Screenshots (if applicable)

N/A

### CODE REVIEW

File: src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts [hunk 1]
```diff
@@ -53,3 +58,110 @@ export function getTradeType(
       return undefined
   }
 }
+
+// TODO: we need to fetch these contracts dynamically, as the whitelist can change
+export enum AggregatorContract {
+  TSAggregatorUniswapV3_100 = '0xbd68cbe6c247e2c3a0e36b8f0e24964914f26ee8',
+  TSAggregatorUniswapV3_500 = '0xe4ddca21881bac219af7f217703db0475d2a9f02',
+  TSAggregatorUniswapV3_3000 = '0x11733abf0cdb43298f7e949c930188451a9a9ef2',
+  TSAggregatorUniswapV3_10000 = '0xb33874810e5395eb49d8bd7e912631db115d5a03',
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:67]: We blackhole funds if any of these are wrong - they must be whitelisted on https://gitlab.com/thorchain/thornode/-/blob/develop/x/thorchain/aggregators/dex_mainnet_current.go

File: src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts [hunk 2]
```diff
@@ -53,3 +58,117 @@ export function getTradeType(
...
+      const poolContract = getContract({
+        abi: IUniswapV3PoolABI,
+        address,
+        publicClient,
+      })
+      try {
+        const [token0, token1]: [Address, Address] = await Promise.all([
+          poolContract.read.token0(),
+          poolContract.read.token1(),
+        ])
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:128]: if we stored the tokens in the result of `generateV3PoolAddressesAcrossFeeRange` such that the index generated is 
```
Map<Address, { tokenA: Address; tokenB: Address; fee: FeeAmount }>
```

.. then we can avoid 2 http requests here

Comment (woodenfurniture) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:128]: oh i see, this is actually required to determine the order of the tokens. Ignore me

File: src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts [hunk 3]
```diff
@@ -53,3 +58,117 @@ export function getTradeType(
...
+  quotedAmounts: Map<Address, bigint>,
+): [Address, bigint] | undefined => {
+  return Array.from(quotedAmounts.entries()).reduce(
+    (
+      addressWithHighestAmount: [Address, bigint] | undefined,
+      [poolAddress, amount]: [Address, bigint],
+    ) => {
+      return amount > (addressWithHighestAmount?.[1] ?? BigInt(0))
+        ? [poolAddress, amount]
+        : addressWithHighestAmount
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:170]:
```suggestion
      if (addressWithHighestAmount === undefined) return [poolAddress, amount]
     
      return amount > addressWithHighestAmount.[1]
        ? [poolAddress, amount]
        : addressWithHighestAmount
```

Comment (woodenfurniture) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:170]: non-blocking of course as they're equivalent it seems

**Review by woodenfurniture — APPROVED** [review]
lgtm, no blocking comments

File: src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts]: Addressed in https://github.com/shapeshift/web/pull/5870/commits/3214c8f434fa895aaa73389850d83f41245609b4

File: src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts (hunk 2 above)
Comment (0xean) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:128]: not a big deal, but all tokens in Uniswap should be sorted alphabetically, so we could avoid the calls. 

see here:
https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/UniswapV3Factory.sol#L41

Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/longTailHelpers.ts:128]: Addressed in https://github.com/shapeshift/web/pull/5902 🙏

**Pull request merged**